pkg/postage/batchstore/reserve_test.go

```diff
@@ -0,0 +1,306 @@
+			}
+		}
+	})
+	t.Run("dilute batches", func(t *testing.T) {
+		n := 0
+		for id := range batches {
+			b, err := bStore.Get([]byte(id))
+			if err != nil {
+				if errors.Is(storage.ErrNotFound, err) {
+					continue
+				}
+				t.Fatal(err)
+			}
+			cs, err := nextChainState(bStore)
+			if err != nil {
+				t.Fatal(err)
+			}
+			if err = increaseDepth(bStore, cs, b); err != nil {
+				t.Fatal(err)
+			}
+			if _, err = checkReserve(bStore, unreserved); err != nil {
+				t.Fatal(err)
+			}
+			n++
+			if n > len(batches)/5 {
+				break
+			}
+		}
+	})
+}
+
+func TestBatchStoreUnreserveAll(t *testing.T) {
+	// temporarily reset reserve Capacity
+	defer func(i int64) {
+		batchstore.Capacity = i
+	}(batchstore.Capacity)
+	batchstore.Capacity = batchstore.Exp2(16)
+
+	bStore, unreserved := setupBatchStore(t)
+	var batches [][]byte
+	// iterate starting from batchstore.DefaultDepth to maxPO
+	_, depth := batchstore.GetReserve(bStore)
+	for step := 0; depth < swarm.MaxPO; step++ {
+		cs, err := nextChainState(bStore)
+		if err != nil {
+			t.Fatal(err)
+		}
+		event := rand.Intn(6)
+		//  0:  dilute, 1: topup, 2,3,4,5: create
+		var b *postage.Batch
+		if event < 2 && len(batches) > 10 {
+			for {
+				n := rand.Intn(len(batches))
+				b, err = bStore.Get(batches[n])
+				if err != nil {
+					if errors.Is(storage.ErrNotFound, err) {
+						continue
+					}
+					t.Fatal(err)
+				}
+				break
+			}
+			if event == 0 {
+				if err = increaseDepth(bStore, cs, b); err != nil {
+					t.Fatal(err)
+				}
+			} else if err = topUp(bStore, cs, b); err != nil {
+				t.Fatal(err)
+			}
+		} else if b, err = createBatch(bStore, cs, depth); err != nil {
+			t.Fatal(err)
+		} else {
+			batches = append(batches, b.ID)
+		}
+		if depth, err = checkReserve(bStore, unreserved); err != nil {
+			t.Fatal(err)
+		}
+	}
+}
+
+func setupBatchStore(t *testing.T) (postage.Storer, map[string]uint8) {
+	t.Helper()
+	// we cannot  use the mock statestore here since the iterator is not giving the right order
+	// must use the leveldb statestore
+	dir, err := ioutil.TempDir("", "batchstore_test")
+	if err != nil {
+		t.Fatal(err)
+	}
+	t.Cleanup(func() {
+		if err := os.RemoveAll(dir); err != nil {
+			t.Fatal(err)
+		}
+	})
+
+	stateStore, err := leveldb.NewStateStore(dir)
+	if err != nil {
+		t.Fatal(err)
+	}
+	t.Cleanup(func() {
+		if err := stateStore.Close(); err != nil {
+			t.Fatal(err)
+		}
+	})
+
+	// set mock unreserve call
+	unreserved := make(map[string]uint8)
+	unreserveFunc := func(batchID []byte, radius uint8) error {
+		unreserved[string(batchID)] = radius
+		return nil
+	}
+	bStore, _ := batchstore.New(stateStore, unreserveFunc)
+
+	// initialise chainstate
+	err = bStore.PutChainState(&postage.ChainState{
+		Block: 666,
+		Total: big.NewInt(0),
+		Price: big.NewInt(1),
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+	return bStore, unreserved
+}
+
+func nextChainState(bStore postage.Storer) (*postage.ChainState, error) {
+	cs := bStore.GetChainState()
```

Comment:
use `t.Helper()` and fail directly in the helper, it will make the tests more readable